### ISSUE TITLE
Improve validation feedback?

### ISSUE THREAD

**Issue by sdruskat** [dieghernan/cff-validator#3]
Hi again,

This is something that all validation using generic JSON Schema validation packages have to deal with: the validation results may not be found very helpful in some cases, especially for users not experienced in schema validation. Improving the output is a lot of work and needs implementation of extra logic, but would help the user. Also, perhaps this could also be done in the upstream dependency itself. So this is more of a suggestion for discussion than a feature request.

E.g., 
- in [this run](https://github.com/sdruskat/key-complete-cff-vaildation/runs/3521972447?check_suite_focus=true), the error message is `1  data has additional properties` (there is an extra key `extra` in the test file). It'd be great if in this case, the name of the addtl. property was provided.
- In [this run](https://github.com/sdruskat/key-complete-cff-vaildation/runs/3522036887?check_suite_focus=true), the error message is more helpful: `1 data["date-released"] referenced schema does not match`. I think this is good enough. :heavy_check_mark: 
- In [this run](https://github.com/sdruskat/key-complete-cff-vaildation/runs/3522070973?check_suite_focus=true), the error is `1 data.authors is required`. Good enough :heavy_check_mark:

Additionally, I'm not sure that the `field` and `message` output are helpful/well-aligned with the output from the validation package:

```bash
------

                  field                          message
1 data["date-released"] referenced schema does not match
```

**Comment by dieghernan** [comment]
I would need to dig in further on this, as I am relying on [jsonvalidate](https://github.com/ropensci/jsonvalidate) R package for validating and I don't have any prior experience on json validating.

A bit off-topic, but you mentioned on https://github.com/citation-file-format/citation-file-format/issues/340:

> I'll take it for a test drive with the key-complete CFF file we have for tests.

Is [this one](https://github.com/sdruskat/key-complete-cff-vaildation/blob/master/1.2.0/pass/key-complete/CITATION.cff)? If so, may I borrow it to complete my tests?

Thanks

**Comment by dieghernan** [comment]
Working on printing suggestion, now the error shows

```
Table: See errors

|field          |message          |
|:--------------|:----------------|
|data.authors.0 |no schemas match |
```

This is pure markdown, so it can be also copied and pasted like this:

Table: See errors

|field          |message          |
|:--------------|:----------------|
|data.authors.0 |no schemas match |

**Comment by sdruskat** [comment]
> I would need to dig in further on this, as I am relying on [jsonvalidate](https://github.com/ropensci/jsonvalidate) R package for validating and I don't have any prior experience on json validating.

Yes, I think it's actually a bit of work, and may not always work as expected, as you'd have to catch the original error messages (which may also not be stable across versions of the validation package dependency), then process them and probably add heuristics to cover the CFF semantics.
 
> A bit off-topic, but you mentioned on [citation-file-format/citation-file-format#340](https://github.com/citation-file-format/citation-file-format/issues/340):
> 
> > I'll take it for a test drive with the key-complete CFF file we have for tests.
> 
> Is [this one](https://github.com/sdruskat/key-complete-cff-vaildation/blob/master/1.2.0/pass/key-complete/CITATION.cff)? If so, may I borrow it to complete my tests?

Yes, it's that one. Go ahead and use it for tests, absolutely.

**Comment by sdruskat** [comment]
> 
> 
> Working on printing suggestion, now the error shows
> 
> ```
> Table: See errors
> 
> |field          |message          |
> |:--------------|:----------------|
> |data.authors.0 |no schemas match |
> ```
> 
> This is pure markdown, so it can be also copied and pasted like this:
> 
> Table: See errors
> field 	message
> data.authors.0 	no schemas match

Nice one! That's definitely an improvement.

**Comment by dieghernan** [comment]
Just for info, what I get from this:

https://github.com/dieghernan/cff-validator/blob/90b5a70f818899bc8560ef919eac7597ef79c576/examples/key-error/CITATION.cff#L1-L19

Is:

Table: **./examples/key-error/CITATION.cff errors:**

|field           |message                          |
|:---------------|:--------------------------------|
|data            |has additional properties        |
|data.authors.0  |no schemas match                 |
|data.doi        |referenced schema does not match |
|data.keywords.0 |is the wrong type                |
|data.license    |referenced schema does not match |
|data.url        |referenced schema does not match |

**Comment by dieghernan** [comment]
I migrated the JSON schema validator to [ajv](https://github.com/ajv-validator/ajv) engine. Now the feedback is much more complete.

## Source file

https://github.com/dieghernan/cff-validator/blob/4cb057ac90c95299dbb9ac38b64764727d50db73/examples/key-error/CITATION.cff#L1-L19

## Before

> :x: ./examples/key-error/CITATION.cff has errors
> 
> |field          |message                          |
> |:--------------|:--------------------------------|
> |cff            |has additional properties        |
> |cff.authors.0  |no schemas match                 |
> |cff.doi        |referenced schema does not match |
> |cff.keywords.0 |is the wrong type                |
> |cff.license    |referenced schema does not match |
> |cff.url        |referenced schema does not match |
> 
> 
> See [Guide to Citation File Format schema version 1.2.0](https://github.com/citation-file-format/citation-file-format/blob/main/schema-guide.md) for debugging.

## With new ajv validator (new implementation)

> :x: ./examples/key-error/CITATION.cff has errors
> 
> |field               |message                                                                              |
> |:-------------------|:------------------------------------------------------------------------------------|
> |cff/                |must NOT have additional properties                                                  |
> |cff/authors/0       |must NOT have additional properties                                                  |
> |cff/authors/0/orcid |must match pattern "https://orcid\.org/[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9X]{1}" |
> |cff/authors/0/orcid |must match format "uri"                                                              |
> |cff/authors/0       |must have required property 'name'                                                   |
> |cff/authors/0       |must NOT have additional properties                                                  |
> |cff/authors/0       |must NOT have additional properties                                                  |
> |cff/authors/0       |must NOT have additional properties                                                  |
> |cff/authors/0/orcid |must match pattern "https://orcid\.org/[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9X]{1}" |
> |cff/authors/0/orcid |must match format "uri"                                                              |
> |cff/authors/0       |must match a schema in anyOf                                                         |
> |cff/doi             |must match pattern "^10\.\d{4,9}(\.\d+)?/[A-Za-z0-9:/_;\-\.\(\)\[\]\\]+$"            |
> |cff/keywords/0      |must be string                                                                       |
> |cff/license         |must be equal to one of the allowed values                                           |
> |cff/license         |must be array                                                                        |
> |cff/license         |must match exactly one schema in oneOf                                               |
> |cff/url             |must match pattern "^(https&#124;http&#124;ftp&#124;sftp)://.+"                      |
> |cff/url             |must match format "uri"                                                              |
> 
> 
> See [Guide to Citation File Format schema version 1.2.0](https://github.com/citation-file-format/citation-file-format/blob/main/schema-guide.md) for debugging.
